README.md
# MacDash

[![Build Status](https://travis-ci.com/skeptycal/macdash.svg?branch=master)](https://travis-ci.com/skeptycal/macdash)
[![Coverage Status](https://coveralls.io/repos/skeptycal/macdash/badge.png?branch=master)](https://coveralls.io/r/skeptycal/macdash?branch=master)

MacDash is a system information web dashboard for macOS using data served by [psutil](https://github.com/giampaolo/psutil) and Flask. It is based on the Linux utility [macdash](https://github.com/Jahaja/macdash).

- [MacDash](#macdash)
  - [Features](#features)
  - [Installation](#installation)
  - [Getting started](#getting-started)
  - [Configuration](#configuration)
  - [Screenshots](#screenshots)
  - [License](#license)

## Features

* **Overview**<br>
  Dashboard overview of the system displaying data on cpu, disks, network, users, memory, swap and network.
* **Processes**<br>
    List processes (`top` like) and view detailed process information about each process.

    Apart from a detailed process overview this is also available for each process:
    * Open files
    * Open connections
    * Memory maps
    * Child processes
    * Resource limits
* **Disks**<br>
    List info on all disks and partitions.
* **Network**<br>
    List info on all network interfaces and the current throughput.
    System-wide open connections listing with filtering. Somewhat like `netstat`.
* **Logs**<br>
    Tail and search logs.
    The logs are added by patterns (like `/var/log/*.log`) which are checked periodically to account for new or deleted files.
* **Multi-node/Cluster**
    Support for multiple agent nodes that is either specified by a config or will register themselves on start-up to a common MacDash node that runs the web interface.
* **All data is updated automatically, no need to refresh**

The GUI is pretty much a modified bootstrap example.
If you got a feel for design and would like to improve the UI of MacDash, please create a pull request with your changes.
It would be much appreciated as there's much room for improvements.

## Installation

Make sure your system is able to build Python C extensions. On Debian derivatives such as Ubuntu this should translate to installing the `build-essential` and `python-dev` packages using `apt-get`:

`# apt-get install build-essential python-dev`

And on RHEL (Fedora, CentOS etc.) distributions:
```
# yum groupinstall "Development Tools"
# yum install python-devel
```

Installation using pip:<br>
`$ pip install macdash`

Since pip [1.5.1](https://github.com/pypa/pip/issues/1423) you are forced to add `--allow-external argparse` since the latest argparse is hosted on google code.

Installation from source:<br>
`$ python setup.py install`

## Getting started

Starting MacDash:<br>
`$ macdash`

Starting a MacDash agent:<br>
`$ macdash -a --register-to [http|https]://[host]:[port] --register-as my-agent-node`

This will start MacDash in *agent mode* and try to register the node to the main MacDash node pointed to by the `--register-to` option.
An agent node will setup an RPC server rather than a webserver at the host and port specified by `-p/--port` and `-b/--bind` respectively.
The main MacDash node (serving HTTP) will present a list of registered nodes that are available to switch between.

Available command-line arguments:
```
$ macdash --help
usage: macdash [-h] [-l path] [-b host] [-p port] [-d] [-a]
              [--register-to host:port] [--register-as name]

macdash [version] - system information web dashboard

optional arguments:
  -h, --help            show this help message and exit
  -l path, --log path   log files to make available for macdash. Patterns (e.g.
                        /var/log/**/*.log) are supported. This option can be
                        used multiple times.
  -b host, --bind host  host to bind to. Defaults to 0.0.0.0 (all interfaces).
  -p port, --port port  port to listen on. Defaults to 5000.
  -d, --debug           enables debug mode.
  -a, --agent           Enables agent mode. This launches a RPC server, using
                        zerorpc, on given bind host and port.
  --register-to host:port
                        The macdash node running in web mode to register this
                        agent to on start up. e.g 10.0.1.22:5000
  --register-as name    The name to register as. (This will default to the
                        node's hostname)
```

## Configuration

macdash uses the configuration handling offered by Flask.
The configuration file pointed to by the environment varible `MACDASH_CONFIG` will be read on startup.<br>
e.g: `$ MACDASH_CONFIG=/home/user/config.py macdash`

In addition to the [built-in configuration values that comes with Flask](http://flask.pocoo.org/docs/config/#builtin-configuration-values) there's a number of macdash specific ones as well:

| Name                               | Description                                                                                                                                                                            |
| ---------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `MACDASH_AUTH_USERNAME`            | When this value and `MACDASH_AUTH_PASSWORD` is set, Basic Authentication will be enabled with the provided credentials. The username of the basic authentication                       |
| `MACDASH_AUTH_PASSWORD`            | The password of the basic authentication                                                                                                                                               |
| `MACDASH_ALLOWED_REMOTE_ADDRESSES` | If this is set, only provided ip addresses will be allowed to access macdash. Addresses is separated by a comma. e.g: `MACDASH_ALLOWED_REMOTE_ADDRESSES = "10.0.0.2, 192.29.20.2"`     |
| `MACDASH_URL_PREFIX`               | This can be used to make macdash serve from a non-root location. e.g: `MACDASH_URL_PREFIX = "/macdash"` would make macdash serve it's pages from /macdash                              |
| `MACDASH_LOG_LEVEL`                | The log level set for macdash (passed in to `logging.basicConfig`). *Defaults to `logging.INFO`*.                                                                                      |
| `MACDASH_LOG_LEVEL`                | The log format set for macdash (passed in to `logging.basicConfig`). *Defaults to `%(levelname)s | %(name)s | %(message)s`*.                                                           |
| `MACDASH_NODES`                    | A list of macdash agent nodes (a dict per node) to register on startup. e.g `[{'name': 'mywebnode', 'host': '10.0.0.2', 'port': 5000}]`                                                |
| `MACDASH_NET_IO_COUNTER_INTERVAL`  | The interval in seconds to update the counters used for calculating network traffic. *Defaults to 3*.                                                                                  |
| `MACDASH_LOGS_INTERVAL`            | The interval in seconds to reapply the log patterns to make sure that file-system changes are applied (log files being created or removed). *Defaults to 60*.                          |
| `MACDASH_REGISTER_INTERVAL`        | The interval in seconds to register the agent to the host macdash node. This is done periodically to be able to determine if any node has gone away and at what time. *Defaults to 60* |
| `MACDASH_LOGS`                     | Log patterns to apply at startup. e.g `['/var/log/*.log']`. To override this option using the command-line use the `-l/--log` arg option.                                              |
| `MACDASH_REGISTER_TO`              | When running in agent mode, this is used to set which macdash node to register the agent node to. e.g `http://10.0.20.2:5000`.                                                         |
| `MACDASH_REGISTER_AS`              | When running in agent mode, this is used to set the name to register as to the host macdash node specified by `MACDASH_REGISTER_TO`.                                                   |
| `MACDASH_HTTPS_KEYFILE`            | Path to the SSL key file to use to enable starting the macdash webserver in HTTPS mode. e.g `/home/user/private.key`                                                                   |
| `MACDASH_HTTPS_CERTFILE`           | Path to the SSL certificate file to use to enable starting the macdash webserver in HTTPS mode. e.g `/home/user/certificate.crt`                                                       |
| `MACDASH_ENVIRON_WHITELIST`        | If set, only the env vars in this list will be displayed with value. e.g `['HOME']`                                                                                                    |

## Screenshots

Overview:
![Overview](docs/screenshots/overview.png)
Listing processes:
![Processes](docs/screenshots/processes.png)
Overview of a process:
![Process overview](docs/screenshots/process_overview.png)
Connections of a process:
![Process connections](docs/screenshots/process_connections.png)
Environment of a process:
![Process environment](docs/screenshots/process_environment.png)
Network:
![Network](docs/screenshots/network.png)
Disks:
![Network](docs/screenshots/disks.png)
Tailing a log:
![Log](docs/screenshots/log.png)
Searching a log:
![Log search](docs/screenshots/log_search.png)

## License
Released under CC0 (Public Domain Dedication).

http://creativecommons.org/publicdomain/zero/1.0/


CHANGELOG
--- v1.0.0 [2019-09-12] ---
* [NEW] Changed a PSDASH_ prefix on all macdash specific configuration options to MACDASH_

* [NEW] Reworked setup.py to change references from psdash to macdash

* [NEW] update classifiers and descriptions to reflect macOS and macdash

* [NEW] update and refactor tests as needed due to dependency update

* [NEW] update dependencies to modern versions to eliminate security issues

* [NEW] dependency management managed by pipenv (automating pip and pyenv)

* [NEW] added [DEV] dependencies autopep8 = "*", pylint = "*"

------------------------------------------------------------------------------
--- Version 1.0.0
--- Copyright (c) 2019 Michael Treanor
--- MacDash <https://github.com/skeptycal/macdash>
--- * Major refactor to change from Linux focus to macOS focus
--- * Addition of many new features, many specific to macOS
--- * Addition of UI design in place of original sample boilerplate
------------------------------------------------------------------------------
--- original license was CC0 (1.0) Universal
--- NO COPYRIGHT
--- psdash <https://github.com/Jahaja/psdash>
--- by Joakim Hamrén <https://github.com/Jahaja>
------------------------------------------------------------------------------

--- v0.6.2 [2015-12-02] ---

* [NEW] Changed gevent version from 1.0.1 to 1.0.2.

* [FIX] Fixed bug that showed incorrect disk usage percentage on the overview page.

--- v0.6.1 [2014-12-13] ---

* [NEW] Changed psutil version from 2.1.1 to 2.1.3

* [FIX] Fixed bug that caused file-based configuration options to be wrongly overriden by defaulted program args.

* [FIX] Added a fix for an unhandled error in psutil when the uid of a process is not associated with an user.

* [NEW] Added a whitelist option to make it possible to decide which env vars values to display.

--- v0.6.0 [2014-09-24] ---

* [NEW] Major updates to the UI design.

* [NEW] Added user proceses filtering to processes page.

* [FIX] Fixed bug where the highlighted text when searching a log could be misplaced.

--- v0.5.0 [2014-09-06] ---

* [NEW] Added HTTPS support.

* [NEW] Added support for multiple hosts.

* [NEW] Changed to using the netifaces lib for reading the host ip-address. (Thanks to Leonardo Santagada)

* [NEW] Made it possible to filter on remote/local address port on the system-wide connections page.

--- v0.4.0 [2014-07-24] ---

* [NEW] Added a environment variables sub-page to the process view.

* [NEW] Updated psutil to 2.1.1

* [NEW] Added a system-wide connections table on the network page.

* [FIX] Removed accidentally committed print statement.

* [NEW] Added more tests and made some refactorings as a consequence.

* [NEW] Added Travis CI setup and test coverage.

* [NEW] Added filters to system-wide connections table

* [NEW] Some rather large refactorings in general due to added tests.

--- v0.3.0 [2014-04-15] ---

* [NEW] Added a PSDASH_URL_PREFIX configuration setting to allow psdash to be served from a sub-url rather than always on root.

* [NEW] Now updating the list of available logs every minute.

* [NEW] The list of logs is now sorted in alphabetical order.

* [FIX] Fixed bug where the view logs page would crash if a previously added log becomes unavailable.

* [FIX] Fix for Python 3 support.

--- v0.2.0 [2014-04-05] ---

* [NEW] It's now possible to pass a pattern (e.g /var/log/**/*.log) when adding log files through the arg option --log.

* [FIX] Fixed issue where file size formatted values would be displayed as zero.

* [FIX] Fixed multiple issues related to non-ascii chars.

* [FIX] Fixed issue where HTML was not escaped in the log contents.

--- v0.1.2 [2014-03-31] ---

* [FIX] Fixed issue where a non-ascii process command-line would cause the process pages to crash.

* [FIX] Fixed issue where network interface info can be defaulted to a empty dict but a key was expected to exist for sorting.

* [FIX] Fixed issue where a network interface alias would cause psdash to crash

* [FIX] Improved the Vagrantfile when using a windows host.

--- v0.1.1 [2014-03-29] ---

* [FIX] Fixed error when specifying a unreadable log file. psdash will now check that
        each log file is readable on startup.

* [FIX] Fixed issue where fetching network interface addresses did not work on 32 bit arch.

* [NEW] Added a PSDASH_ prefix on all psdash specific configuration options.

* [FIX] Added argparse to setup.py requirements for Python 2.6 compatibility


macdash/run.py
from macdash.web import fromtimestamp
from macdash.node import LocalNode, RemoteNode
from macdash import __version__
import zerorpc
from flask import Flask
from logging import getLogger
import urllib2
import urllib
import socket
import logging
import argparse
import locale
from gevent.pywsgi import WSGIServer
import gevent
from gevent.monkey import patch_all
patch_all()


logger = getLogger('macdash.run')


class MacDashRunner(object):
    DEFAULT_LOG_INTERVAL = 60
    DEFAULT_NET_IO_COUNTER_INTERVAL = 3
    DEFAULT_REGISTER_INTERVAL = 60
    DEFAULT_BIND_HOST = '0.0.0.0'
    DEFAULT_PORT = 5000
    LOCAL_NODE = 'localhost'

    @classmethod
    def create_from_cli_args(cls):
        return cls(args=None)

    def __init__(self, config_overrides=None, args=tuple()):
        self._nodes = {}
        config = self._load_args_config(args)
        if config_overrides:
            config.update(config_overrides)
        self.app = self._create_app(config)

        self._setup_nodes()
        self._setup_logging()
        self._setup_context()

    def _get_args(self, args):
        parser = argparse.ArgumentParser(
            description='macdash %s - system information web dashboard' % __version__
        )
        parser.add_argument(
            '-l', '--log',
            action='append',
            dest='logs',
            default=None,
            metavar='path',
            help='log files to make available for macdash. Patterns (e.g. /var/log/**/*.log) are supported. '
                 'This option can be used multiple times.'
        )
        parser.add_argument(
            '-b', '--bind',
            action='store',
            dest='bind_host',
            default=None,
            metavar='host',
            help='host to bind to. Defaults to 0.0.0.0 (all interfaces).'
        )
        parser.add_argument(
            '-p', '--port',
            action='store',
            type=int,
            dest='port',
            default=None,
            metavar='port',
            help='port to listen on. Defaults to 5000.'
        )
        parser.add_argument(
            '-d', '--debug',
            action='store_true',
            dest='debug',
            help='enables debug mode.'
        )
        parser.add_argument(
            '-a', '--agent',
            action='store_true',
            dest='agent',
            help='Enables agent mode. This launches a RPC server, using zerorpc, on given bind host and port.'
        )
        parser.add_argument(
            '--register-to',
            action='store',
            dest='register_to',
            default=None,
            metavar='host:port',
            help='The macdash node running in web mode to register this agent to on start up. e.g 10.0.1.22:5000'
        )
        parser.add_argument(
            '--register-as',
            action='store',
            dest='register_as',
            default=None,
            metavar='name',
            help='The name to register as. (This will default to the node\'s hostname)'
        )

        return parser.parse_args(args)

    def _load_args_config(self, args):
        config = {}
        for k, v in vars(self._get_args(args)).iteritems():
            if v:
                key = 'MACDASH%s' % k.upper() if k != 'debug' else 'DEBUG'
                config[key] = v
        return config

    def _setup_nodes(self):
        self.add_node(LocalNode())

        nodes = self.app.config.get('MACDASH_NODES', [])
        logger.info("Registering %d nodes", len(nodes))
        for n in nodes:
            self.register_node(n['name'], n['host'], int(n['port']))

    def add_node(self, node):
        self._nodes[node.get_id()] = node

    def get_local_node(self):
        return self._nodes.get(self.LOCAL_NODE)

    def get_node(self, name):
        return self._nodes.get(name)

    def get_nodes(self):
        return self._nodes

    def register_node(self, name, host, port):
        n = RemoteNode(name, host, port)
        node = self.get_node(n.get_id())
        if node:
            n = node
            logger.debug("Updating registered node %s", n.get_id())
        else:
            logger.info("Registering %s", n.get_id())
        n.update_last_registered()
        self.add_node(n)
        return n

    def _create_app(self, config=None):
        app = Flask(__name__)
        app.macdash = self
        app.config.from_envvar('MACDASH_CONFIG', silent=True)

        if config and isinstance(config, dict):
            app.config.update(config)

        self._load_allowed_remote_addresses(app)

        # If the secret key is not read from the config just set it to something.
        if not app.secret_key:
            app.secret_key = 'whatisthissourcery'
        app.add_template_filter(fromtimestamp)

        from macdash.web import webapp
        prefix = app.config.get('MACDASH_URL_PREFIX')
        if prefix:
            prefix = '/' + prefix.strip('/')
        webapp.url_prefix = prefix
        app.register_blueprint(webapp)

        return app

    def _load_allowed_remote_addresses(self, app):
        key = 'MACDASH_ALLOWED_REMOTE_ADDRESSES'
        addrs = app.config.get(key)
        if not addrs:
            return

        if isinstance(addrs, (str, bytes)):
            app.config[key] = [a.strip() for a in addrs.split(',')]

    def _setup_logging(self):
        level = self.app.config.get(
            'MACDASH_LOG_LEVEL', logging.INFO) if not self.app.debug else logging.DEBUG
        format = self.app.config.get(
            'MACDASH_LOG_FORMAT', '%(levelname)s | %(name)s | %(message)s')

        logging.basicConfig(
            level=level,
            format=format
        )
        logging.getLogger('werkzeug').setLevel(
            logging.WARNING if not self.app.debug else logging.DEBUG)

    def _setup_workers(self):
        net_io_interval = self.app.config.get(
            'MACDASH_NET_IO_COUNTER_INTERVAL', self.DEFAULT_NET_IO_COUNTER_INTERVAL)
        gevent.spawn_later(
            net_io_interval, self._net_io_counters_worker, net_io_interval)

        if 'MACDASH_LOGS' in self.app.config:
            logs_interval = self.app.config.get(
                'MACDASH_LOGS_INTERVAL', self.DEFAULT_LOG_INTERVAL)
            gevent.spawn_later(logs_interval, self._logs_worker, logs_interval)

        if self.app.config.get('MACDASH_AGENT'):
            register_interval = self.app.config.get(
                'MACDASH_REGISTER_INTERVAL', self.DEFAULT_REGISTER_INTERVAL)
            gevent.spawn_later(register_interval,
                               self._register_agent_worker, register_interval)

    def _setup_locale(self):
        # This set locale to the user default (usually controlled by the LANG env var)
        locale.setlocale(locale.LC_ALL, '')

    def _setup_context(self):
        self.get_local_node().net_io_counters.update()
        if 'MACDASH_LOGS' in self.app.config:
            self.get_local_node().logs.add_patterns(
                self.app.config['MACDASH_LOGS'])

    def _logs_worker(self, sleep_interval):
        while True:
            logger.debug("Reloading logs...")
            self.get_local_node().logs.add_patterns(
                self.app.config['MACDASH_LOGS'])
            gevent.sleep(sleep_interval)

    def _register_agent_worker(self, sleep_interval):
        while True:
            logger.debug("Registering agent...")
            self._register_agent()
            gevent.sleep(sleep_interval)

    def _net_io_counters_worker(self, sleep_interval):
        while True:
            logger.debug("Updating net io counters...")
            self.get_local_node().net_io_counters.update()
            gevent.sleep(sleep_interval)

    def _register_agent(self):
        register_name = self.app.config.get('MACDASH_REGISTER_AS')
        if not register_name:
            register_name = socket.gethostname()

        url_args = {
            'name': register_name,
            'port': self.app.config.get('MACDASH_PORT', self.DEFAULT_PORT),
        }
        register_url = '%s/register?%s' % (
            self.app.config['MACDASH_REGISTER_TO'], urllib.urlencode(url_args))

        if 'MACDASH_AUTH_USERNAME' in self.app.config and 'MACDASH_AUTH_PASSWORD' in self.app.config:
            auth_handler = urllib2.HTTPBasicAuthHandler()
            auth_handler.add_password(
                realm='MacDash login required',
                uri=register_url,
                user=self.app.config['MACDASH_AUTH_USERNAME'],
                passwd=self.app.config['MACDASH_AUTH_PASSWORD']
            )
            opener = urllib2.build_opener(auth_handler)
            urllib2.install_opener(opener)

        try:
            urllib2.urlopen(register_url)
        except urllib2.HTTPError as e:
            logger.error('Failed to register agent to "%s": %s',
                         register_url, e)

    def _run_rpc(self):
        logger.info("Starting RPC server (agent mode)")

        if 'MACDASH_REGISTER_TO' in self.app.config:
            self._register_agent()

        service = self.get_local_node().get_service()
        self.server = zerorpc.Server(service)
        self.server.bind('tcp://%s:%s' % (self.app.config.get('MACDASH_BIND_HOST', self.DEFAULT_BIND_HOST),
                                          self.app.config.get('MACDASH_PORT', self.DEFAULT_PORT)))
        self.server.run()

    def _run_web(self):
        logger.info("Starting web server")
        log = 'default' if self.app.debug else None

        ssl_args = {}
        if self.app.config.get('MACDASH_HTTPS_KEYFILE') and self.app.config.get('MACDASH_HTTPS_CERTFILE'):
            ssl_args = {
                'keyfile': self.app.config.get('MACDASH_HTTPS_KEYFILE'),
                'certfile': self.app.config.get('MACDASH_HTTPS_CERTFILE')
            }

        listen_to = (
            self.app.config.get('MACDASH_BIND_HOST', self.DEFAULT_BIND_HOST),
            self.app.config.get('MACDASH_PORT', self.DEFAULT_PORT)
        )
        self.server = WSGIServer(
            listen_to,
            application=self.app,
            log=log,
            **ssl_args
        )
        self.server.serve_forever()

    def run(self):
        logger.info('Starting macdash v%s' % __version__)

        self._setup_locale()
        self._setup_workers()

        logger.info('Listening on %s:%s',
                    self.app.config.get('MACDASH_BIND_HOST',
                                        self.DEFAULT_BIND_HOST),
                    self.app.config.get('MACDASH_PORT', self.DEFAULT_PORT))

        if self.app.config.get('MACDASH_AGENT'):
            return self._run_rpc()
        else:
            return self._run_web()


def main():
    r = MacDashRunner.create_from_cli_args()
    r.run()


if __name__ == '__main__':
    main()
